### PR TITLE
extend c++ compatibility to whole libxmp library

### PR DESCRIPTION
**Pull request by sezero**
This is an attempt at extending the C++ compatibility work to whole
library after #419. Only tested with gcc, yet.

The only remaining errors from gcc are in xz depacker which I don't
know whether I want to fix, yet:
```
In file included from src/depackers/xz_dec_lzma2.c:12:
src/depackers/xz_lzma2.h: In function 'void lzma_state_literal(lzma_state*)':
src/depackers/xz_lzma2.h:69: error: no match for 'operator-=' in '* state -= 3'
src/depackers/xz_lzma2.h:71: error: no match for 'operator-=' in '* state -= 6'

src/depackers/xz_dec_stream.c: In function 'xz_ret dec_stream_header(xz_dec*)':
src/depackers/xz_dec_stream.c:410: error: invalid conversion from 'uint8' to 'xz_check'
```
Please review. @Clownacy: please test, if you can.

### CODE REVIEW

**Comment by sezero** [comment]
Huh, forgot that I'll have to fix the linkage issue (c.f. https://github.com/libxmp/libxmp/commit/002ac4175a84a292d8298248bdc032ababf4c73a)

**Comment by sezero** [comment]
Linkage is now fixed when using c++ compilers. All there's left is the xz 'enum' things.

**Comment by sezero** [comment]
The xz depacker now builds in c++ mode, too. All seems done.

**Comment by Clownacy** [comment]
I had to change the `depacker` struct in `depackers/depacker.h` to this in order to get libxmp to compile with my MSVC 6 setup, but other than that, this works great!

```cpp
struct depacker {
	int (*test)(unsigned char *);
	int (*depack)(FILE *, FILE *);
};
```

**Comment by sezero** [comment]
> I had to change the `depacker` struct in `depackers/depacker.h` to this 

Oops, confirmed and now fixed,

> But, other than that, this works great!

Great, thanks for testing!

This is ready for review & merge.

File: src/depackers/inflate.c [hunk 1]
```diff
@@ -1040,16 +1039,14 @@ int libxmp_inflate(FILE *in, FILE *out, uint32 *checksum, int is_zip)
 
   data.huffman_tree_len_static = NULL;
 
-  huffman_tree_len=malloc(HUFFMAN_TREE_SIZE * sizeof(struct huffman_tree_t));
+  huffman_tree_len=(struct huffman_tree_t *) malloc(HUFFMAN_TREE_SIZE * sizeof(struct huffman_tree_t));
   if (huffman_tree_len == NULL)
     goto err;
-
```
Comment (AliceLR) [inline src/depackers/inflate.c:1046]: See previous comment.

File: src/depackers/inflate.c [hunk 2]
```diff
@@ -1040,16 +1039,14 @@ int libxmp_inflate(FILE *in, FILE *out, uint32 *checksum, int is_zip)
...
   memset(huffman_tree_len, 0xff, HUFFMAN_TREE_SIZE * sizeof(struct huffman_tree_t));
 
-  huffman_tree_dist=malloc(HUFFMAN_TREE_SIZE * sizeof(struct huffman_tree_t));
+  huffman_tree_dist=(struct huffman_tree_t *) malloc(HUFFMAN_TREE_SIZE * sizeof(struct huffman_tree_t));
   if (huffman_tree_dist == NULL)
     goto err;
-
```
Comment (AliceLR) [inline src/depackers/inflate.c:1052]: See previous comment.

File: src/hmn_extras.c [hunk 3]
```diff
@@ -88,7 +88,6 @@ int libxmp_hmn_new_instrument_extras(struct xmp_instrument *xxi)
 	if (xxi->extra == NULL)
 		return -1;
 	HMN_INSTRUMENT_EXTRAS((*xxi))->magic = HMN_EXTRAS_MAGIC;
-
```
Comment (AliceLR) [inline src/hmn_extras.c:91]: See previous comments.

File: src/loaders/iff.c [hunk 4]
```diff
@@ -146,10 +146,8 @@ int libxmp_iff_load(iff_handle opaque, struct module_data *m, HIO_HANDLE *f, voi
 
 	while (!hio_eof(f)) {
 		ret = iff_chunk(opaque, m, f, parm);
-
```
Comment (AliceLR) [inline src/loaders/iff.c:149]: See previous comments.

File: src/mix_all.c [hunk 5]
```diff
@@ -170,7 +170,6 @@
     int old_vr = vi->old_vr
 
 #ifndef LIBXMP_CORE_DISABLE_IT
-
```
Comment (AliceLR) [inline src/mix_all.c:173]: See above comments.

File: src/win32/ptpopen.c [hunk 6]
```diff
@@ -70,8 +70,7 @@ static int my_pipe(HANDLE *readwrite)
   sa.bInheritHandle = 1;            /* the child must inherit these handles */
   sa.lpSecurityDescriptor = NULL;
 
-  if (! CreatePipe (&readwrite[0],&readwrite[1],&sa,1 << 13))
-  {
+  if (!CreatePipe(&readwrite[0],&readwrite[1],&sa,1 << 13)) {
```
Comment (AliceLR) [inline src/win32/ptpopen.c:73]: Considering this is contrib code, I don't see the point of (in)correcting its brace style to match the rest of libxmp.

**Review by AliceLR — COMMENTED** [review]
The changes relevant to C++ compatibility look good, but there are a lot of unrelated whitespace changes that don't really help code readability and IMO in several cases do the opposite. I put a comment on each affected file but didn't comment them all (to avoid spam...).

edit: As long as I add `-Wno-sign-compare`, this branch builds fine with g++ without any other warnings.

**Comment by AliceLR** [comment]
Not necessarily within the scope of this branch: should we replace the `#ifdef __cplusplus` etc. with proper `LIBXMP_BEGIN_DECLS`/`LIBXMP_END_DECLS` defines and wrap all of the headers with them?

**Comment by sezero** [comment]
> Not necessarily within the scope of this branch: should we replace the `#ifdef __cplusplus` etc. with proper `LIBXMP_BEGIN_DECLS`/`LIBXMP_END_DECLS` defines and wrap all of the headers with them?

That's a good idea.  Will revise shortly.

File: src/control.c [hunk 7]
```diff
@@ -320,14 +320,12 @@ LIBXMP_EXPORT_VERSIONED extern int xmp_set_player_v43__(xmp_context, int, int)
 			__attribute__((alias("xmp_set_player_v40__"))) LIBXMP_ATTRIB_SYMVER("xmp_set_player@XMP_4.3");
 LIBXMP_EXPORT_VERSIONED extern int xmp_set_player_v44__(xmp_context, int, int)
 			__attribute__((alias("xmp_set_player_v40__"))) LIBXMP_ATTRIB_SYMVER("xmp_set_player@@XMP_4.4");
-
```
[the remark this answers is not in the document]
Comment (sezero) [reply src/control.c:323]: Will be removed in next force-push

File: src/depackers/inflate.c (hunk 1 above)
Comment (sezero) [reply src/depackers/inflate.c:1046]: There is no readability issue with this, it just groups allocation+memset together

File: src/hmn_extras.c (hunk 3 above)
Comment (sezero) [reply src/hmn_extras.c:91]: I see no issues with this: all other lines in these tiny procs have no newlines between them.

File: src/loaders/iff.c (hunk 4 above)
Comment (sezero) [reply src/loaders/iff.c:149]: This version is more readable to me btw..

File: src/mix_all.c (hunk 5 above)
Comment (sezero) [reply src/mix_all.c:173]: Dropping changes to mix_all.c in next force-push

File: src/win32/ptpopen.c (hunk 6 above)
Comment (sezero) [reply src/win32/ptpopen.c:73]: OK, dropping them in next force-push

File: src/depackers/inflate.c (hunk 1 above)
Comment (sezero) [reply src/depackers/inflate.c:1046]: kept this one.

File: src/hmn_extras.c (hunk 3 above)
Comment (sezero) [reply src/hmn_extras.c:91]: kept this one.

File: src/loaders/iff.c (hunk 4 above)
Comment (sezero) [reply src/loaders/iff.c:149]: kept this one.

**Comment by sezero** [comment]
Revised version force-pushed.

**Comment by sezero** [comment]
> edit: As long as I add `-Wno-sign-compare`, this branch builds fine with g++ without any other warnings.

Yep. `-Wno-sign-compare` can be handled later, if we _truly_ care about it

>  with clang++, aside from it issuing `-Wdeprecated` complaints for building C with clang++,

That's clang for you..

> as well as complaints about usage of `register`

IIRC, c++11 (or was it 17? whatever) version dropped support for `register`

> (why is libxmp even using `register`?).

Possibly as a hint to compiler? Not sure.

**Review by AliceLR — APPROVED** [review]
This looks good to me.

**Pull request merged**